qiskit/transpiler/target.py

```diff
@@ -600,6 +606,53 @@ def physical_qubits(self):
         """Returns a sorted list of physical_qubits"""
         return list(range(self.num_qubits))
 
+    def get_non_global_operation_names(self, strict_direction=False):
+        """Return the non-global operation names for the target
+
+        The non-global operations are those in the target which don't apply
+        on all qubits (for single qubit operations) or all multiqubit qargs
+        (for multi-qubit operations).
+
+        Args:
+            strict_direction (bool): If set to ``True`` the multi-qubit
+                operations considered as non-global respect the strict
+                direction (or order of qubits in the qargs is signifcant). For
+                example, if ``cx`` is defined on ``(0, 1)`` and ``ecr`` is
+                defined over ``(1, 0)`` by default neither would be considered
+                non-global, but if ``strict_direction`` is set ``True`` both
+                ``cx`` and ``ecr`` would be returned.
+
+        Returns:
+            List[str]: A list of operation names for operations that aren't global in this target
+        """
+        if strict_direction:
+            if self._non_global_strict_basis is not None:
+                return self._non_global_strict_basis
+            search_set = self._qarg_gate_map.keys()
+        else:
+            if self._non_global_basis is not None:
+                return self._non_global_basis
+
+            search_set = {frozenset(qarg) for qarg in self._qarg_gate_map if len(qarg) != 1}
+        incomplete_basis_gates = []
+        size_dict = defaultdict(int)
+        size_dict[1] = self.num_qubits
+        for qarg in search_set:
+            if len(qarg) == 1:
+                continue
+            size_dict[len(qarg)] += 1
+        for inst, qargs in self._gate_map.items():
+            qarg_sample = next(iter(qargs))
+            if qarg_sample is None:
+                continue
+            if len(qargs) != size_dict[len(qarg_sample)]:
```

Comment:
I might be wrong here but I feel this should be:
```suggestion
            if len(qargs) < size_dict[len(qarg_sample)]:
```
since currently the following returns:
```python
gmap = Target()
cx_props = {
    (0, 1): InstructionProperties(duration=5.23e-7, error=0.00098115),
    (1, 0): InstructionProperties(duration=4.52e-7, error=0.00132115),
}
gmap.add_instruction(CXGate(), cx_props)
gmap.get_non_global_operation_names(strict_direction=False)
---
['cx']
```

Reply:
Ah, yep good catch. We need to use the set to treat it as a weak edge otherwise we're comparing number of strong edges for the operation to the number of weak edges in the global target which doesn't make sense.

Reply:
Well making this change breaks the test I added for the non-local 1q gates. It's the same bug we had before where it's not including non-local subsets in the search target because cx is now global in that target with `strict_direction=False`. I'll have to adjust that logic to work over all operations >=2q not just the non-local ones.

Reply:
ok I think I've fixed this now here: https://github.com/Qiskit/qiskit-terra/pull/7227/commits/480159481f396e87df1d3e68e146499cbca19ad9